pkg/cached/cache.go

```diff
@@ -0,0 +1,257 @@
+//     error pops-up.
+//
+// # Atomicity
+//
+// Most of the operations are not atomic/thread-safe, except for
+// [Replaceable.Replace] which can be performed while the objects
+// are being read.
+//
+// # Etags
+//
+// Etags in this library is a cache version identifier. It doesn't
+// necessarily strictly match to the semantics of http `etags`, but are
+// somewhat inspired from it and function with the same principles.
+// Hashing the content is a good way to guarantee that your function is
+// never going to be called spuriously. In Kubernetes world, this could
+// be a `resourceVersion`, this can be an actual etag, a hash, a UUID
+// (if the cache always changes), or even a made-up string when the
+// content of the cache never changes.
+package cached
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// Result is the content returned from a call to a cache. It can either
+// be created with [NewResultOK] if the call was a success, or
+// [NewResultErr] if the call resulted in an error.
+type Result[T any] struct {
+	Data T
+	Etag string
+	Err  error
+}
+
+// NewResultData creates a new [Result] for a successful operation.
+func NewResultOk[T any](data T, etag string) Result[T] {
+	return Result[T]{
+		Data: data,
+		Etag: etag,
+	}
+}
+
+// NewResultErr creates a new [Result] when an error has happened.
+func NewResultErr[T any](err error) Result[T] {
+	return Result[T]{
+		Err: err,
+	}
+}
+
+// Result can be treated as a [Data] if necessary.
+func (r Result[T]) Get() Result[T] {
+	return r
+}
+
+// Data is a cache that performs an action whose result data will be
+// cached. It also returns an "etag" identifier to version the cache, so
+// that the caller can know if they have the most recent version of the
+// cache (and can decide to cache some operation based on that).
+//
+// The [NewMerger] and [NewTransformer] automatically handle
+// that for you by checking if the etag is updated before calling the
+// merging or transforming function.
+type Data[T any] interface {
+	// Returns the cached data, as well as an "etag" to identify the
+	// version of the cache, or an error if something happened.
+	Get() Result[T]
+}
+
+// T is the source type, V is the destination type.
+type merger[K comparable, T, V any] struct {
+	mergeFn      func(map[K]Result[T]) Result[V]
+	caches       map[K]Data[T]
+	cacheResults map[K]Result[T]
+	result       Result[V]
+}
+
+// NewMerger creates a new merge cache, a cache that merges the result
+// of other caches. The function only gets called if any of the
+// dependency has changed.
+//
+// If any of the dependency returned an error before, or any of the
+// dependency returned an error this time, or if the mergeFn failed
+// before, then the function is reran.
+//
+// The caches and results are mapped by string so that associated data
+// can be retrieved.
+func NewMerger[K comparable, T, V any](mergeFn func(results map[K]Result[T]) Result[V], caches map[K]Data[T]) Data[V] {
```

Comment:
is it safe for users to insert or remove map items after creating this? Maybe explain in the comment?